---
title: "salmon CytB taxonomic filtering"
output: html_notebook
---

15 February 2022

This should be all salmon all of the time...


```{r load-libraries}
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)

```

```{r load-output-from-taxonkit}
pool_taxonomy <-read.delim("blast_outputs/CytB_96perc_taxlineageV2.txt", header = FALSE)

head(pool_taxonomy)
```


# clean up the header a bit

```{r}
# use the full taxonomy rather than the seq id to collapse identical entries
pool_tax_df <- pool_taxonomy %>%
  filter(V4 > 100) %>% # make sure all retained matches are >100 bp
  select(-V2, -V5, -V6, -V7, -V14, -V8, -V9, -V10, -V11, -V12, -V15, -V16) %>%  #remove unnecessary columns
  group_by(V1, V17) %>% # group by the sequence key and the full taxonomy to reduce duplicate entries
  unique() %>% # doing that reduced the number of entries from 146k to 17k
  rename(qseqid=V1, perc_id=V3, length=V4, sscinames=V13, taxonomy=V17) %>% #rename headers
  filter(!str_detect(sscinames, "environmental")) %>% # filter out any environmental samples
  filter(!str_detect(sscinames, "synthetic")) %>% # filter out any synthetic "samples"
  filter(perc_id >= 98) # seems like some of the matches below 98% are dubious (jellyfish and herring <1% different??)
 
```

```{r how-many-total-ASVs-w-taxonomy}
pool_tax_df %>% 
  ungroup() %>%
  select(qseqid) %>%
  unique()
```


Maybe break up that taxonomic info so that we can focus on the things that we can identify to some relevant level (at least order!?)
```{r}
# first figure out where there are issues with that taxonomy format:
genus.hybrids <- pool_tax_df %>%
  filter(!str_detect(taxonomy, ";")) %>% 
  # looking at these, they're probably hybrids or not good species, which is why the taxonomy broke down. 
  separate(sscinames, into = c("genus"), extra = "drop", remove = FALSE)

# and where there aren't issues
taxon_df <- pool_tax_df %>%
  filter(str_detect(taxonomy, ";")) %>%
  separate(taxonomy, into=c("kingdom", "phylum", "class", "order", "family", "genus", "species"), sep = ";")

# create a separate taxonomic lineage df with unique entries
lineage_df <- taxon_df %>%
  ungroup() %>%
  select(kingdom, phylum, class, order, family, genus) %>%
  unique()


# there are only 18 unique genera in this small messed up part of the dataset 
genus_fixed <- genus.hybrids %>%
  ungroup() %>%
  select(genus) %>%
  unique() %>% # can I fix the lineage by adding existing data?
  left_join(lineage_df, by = "genus") %>% # now the Tachyeres duck is the only loner
# fix that manually
  mutate(family = ifelse(genus == "Tachyeres", 
                         "Anatidae", family)) %>%
    mutate(order = ifelse(genus == "Tachyeres",
                        "Anseriformes", order)) %>%
      mutate(class = ifelse(genus == "Tachyeres",
                        "Aves", class)) %>%
        mutate(phylum = ifelse(genus == "Tachyeres", 
                               "Chordata", phylum)) %>%
          mutate(kingdom = ifelse(genus == "Tachyeres",
                                    "Eukaryota", kingdom))

# add the lineage info back to the rest of the data
genus_fixed_df <- genus.hybrids %>%
  ungroup() %>%
  left_join(genus_fixed, by = "genus") %>%
  select(-taxonomy)
```
I'm not sure what this is supposed to mean as a scientific name: Clupea harengus;Xystreurys rasile - if it's a hybrid, it seems like the Clupea is more likely because of distribution. 



```{r}
# and merge them together
fixed_tax_df <- taxon_df %>%
  bind_rows(genus_fixed_df) %>%
  # and fix the species syntax with underscores
   mutate_if(is.character, str_replace_all, ' ', '_') %>%
   filter(!str_detect(sscinames, "_x_"))
  # remove hybrids because the mtDNA sequence actually is only one or the other of the parent species

```


## Sort and clean-up the results based on % identity

```{r total-n-seqs}
# how many total sequences?
fixed_tax_df %>%
  select(qseqid) %>%
  unique()

```

There are four categories:
1. sequences that match a single species unambiguously (the minority)

Sequences that match multiple species are divided in three categories:
2. top matches > 2% identity than second-ranked matches
3. top matches < 2% identity than second-ranked matches
4. Multiple top matches with the same % identity

```{r}
# 1. sequences that are unambiguously a single species
single_spp_seqs <- fixed_tax_df %>% 
  group_by(qseqid) %>%
  add_tally(name = "n_taxa") %>%
  filter(n_taxa == 1)
  
```
16 sequences are single-species


```{r}
# remove the single-species seqs from the dataframe and then rank the hits by % identity for the remaining seqs
seq_id_diff <- fixed_tax_df %>%
  anti_join(., single_spp_seqs) %>%
  select(-length) %>%
  group_by(qseqid, species, genus, family, order, class, phylum, kingdom) %>%
    mutate(seq_percID = max(perc_id)) %>%
    group_by(qseqid, species, genus, family, order, class, phylum, kingdom, seq_percID) %>%
  summarise(max(seq_percID)) %>% # take just the top hit for each taxon (for each sequence)
  select(-`max(seq_percID)`) %>%
  ungroup() %>%
  group_by(qseqid) %>%
      mutate(id_rank = rank(desc(seq_percID), ties.method = "min")) %>% # rank the taxonomic hits per sequence by % id
       mutate(top_perc = max(seq_percID)) %>% # designate the highest % id for the best taxonomic hit in each sequence (in some, but not all cases, this is 100%)   
      mutate(diff = top_perc - seq_percID) %>% # calculate the difference between the % identity of the top hit and each subsequent taxonomic hit
        arrange(diff)

seq_id_diff %>%
  filter(diff > 0)

```

Now I have the single best entry for each species for each sequence ranked and with the difference between the first and second ranked entries calculated.

For sequences with multiple top hits, where the difference between ranked taxa = 0, I will end up defaulting to genus- or family-level ID (or carrying the individual species info around in some capacity). I will do the same for any sequences where the difference betweeen the first and second ranked taxa is < 2%.


Figure out which differences are > 2% and eliminate those first?
```{r}
# filter out any taxa that are >2% less matching identity than the top taxonomic hit for a given sequence
to_remove_low_perc_hits <- seq_id_diff %>%
  ungroup() %>%
  group_by(qseqid) %>%
  filter(diff > 2)

keepers <- seq_id_diff %>%
  anti_join(to_remove_low_perc_hits)

```

```{r}
# this data frame includes only those taxonomic hits that should be considered.
# so now I need to determine whether they should be assigned to genus, family, order, etc. 
singletons <- keepers %>%
  select(qseqid) %>%
  tally() %>%
  filter(n == 1)

# these are the seqs that now have only a single match
singleton_df <- singletons %>%
  left_join(keepers) %>%
  select(-n) %>%
  bind_rows(single_spp_seqs) # combine the single spp data

## Genus-level matches
# remove the singletons from the bigger df 
single_genus <- keepers %>%
  anti_join(singleton_df)%>% # at best, these should be genus-level matches
  group_by(qseqid, genus) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single genus
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "genus") %>%
  mutate(taxon = genus)
  

## Family-level matches
single_family <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  group_by(qseqid, family) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single family
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "family") %>%
  mutate(taxon = family)


## Order-level matches
single_order <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  group_by(qseqid, order) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single order
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "order") %>%
  mutate(taxon = order)


## Class-level matches
single_class <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  group_by(qseqid, class) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single class
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "class") %>%
  mutate(taxon = class)


## Phylum-level matches
single_phylum <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, phylum) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single phylum
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "phylum") %>%
  mutate(taxon = phylum)

# Kingdom level matches
single_kingdom <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, kingdom) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single kingdom
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "kingdom") %>%
  mutate(taxon = kingdom)
```

Seems... unlikely that a jellyfish would be 1% dissimilar from a herring. BUT, it's below 98%, so maybe that's at least one reason to be more stringent on the upstream side of this?

Adding that 98% filter in back up top.

Modify the singleton_df to include the right variable headers
```{r}
single_spp <- singleton_df %>%
  select(-perc_id, -length, -sscinames, -n_taxa) %>%
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)

```




```{r}
# recombine the full data set now that the appropriate level of taxonomy has been determined
sorted_tax_df <- bind_rows(single_kingdom, single_phylum, single_class, single_order, single_family, single_genus, single_spp)

# just fish
just.fish <- sorted_tax_df %>%
  filter(class == "Actinopteri") %>% # if I filter by class, I can remove all taxonomic levels above that
  filter(!taxonomic_level %in% c("kingdom", "phylum", "class"))


edna.fish.family <- just.fish %>%
  select(family) %>%
  unique() %>%
  arrange(family)

# take a quick look at the genera included too
just.fish %>%
  select(genus) %>%
  unique() %>%
  arrange(genus)
```

I'm not going to filter by the Fish Atlas because I found that to be marginally helpful and mostly a pain.


```{r}
uncollapsed_taxonomy_cytb <- sorted_tax_df %>%
  select(-kingdom, -phylum, -class, -top_perc, -id_rank)
  

uncollapsed_taxonomy_cytb %>%
  select(qseqid, taxon, taxonomic_level) %>%
  unique()
```

Save the output
```{r}
uncollapsed_taxonomy_cytb %>%
  write_csv("csv_outputs/salmonidCytbTaxonomyDataframe.csv")
```

Go to `12-salmonid-cytb-analysis.Rmd`



##### STOP HERE ##### 
Clinocottus_embryum	 = calico sculpin

Let's bung together the sample data frame to see if the distribution of different salmonid species makes sense and whether read counts seem accurate.


## Read in ASV-sample table for Poolseq data

```{r ASV-sample-table}
asv_tbl <- read.csv("csv_outputs/CytBsampleTable.csv")

# rename the samples because the sample sheet for the MiSeq wasn't accurate (bec of the extra positive controls for salmon species)
# read in the renaming table
library(readxl)

rename_plate1 <- read_xlsx("../cytb/Salmon CytB_Oct2021_updated.xlsx", sheet = "ctyb_plate1_samplesheet")
rename_plate2 <- read_xlsx("../cytb/Salmon CytB_Oct2021_updated.xlsx", sheet = "cytb_plate2_samplesheet")

rename_all <- bind_rows(rename_plate1, rename_plate2) %>%
  mutate(sample = str_replace_all(sample, "_", "-"))

asv_renamed_tbl <- asv_tbl %>%
  left_join(., rename_all, by = c("X"="sample"))

asv_renamed2 <- asv_renamed_tbl %>%
  select(real_sample_name, X, 2:length(asv_renamed_tbl)) %>%
  select(-X) %>%
   rename(sample.PCR = real_sample_name) %>%
    separate(sample.PCR, into = c("sample", "rep", "extra"), remove = F) 

```

Combine taxonomy and sample ASV table
```{r merge-taxonomy-and-sample-tbl}
# pivot longer for ease of joining
asv_long_tbl <- asv_renamed2 %>%
  pivot_longer(cols = 5:length(asv_renamed2), names_to = "ASV", values_to = "count")

# fix the ASV vs. Seq issue
# un_taxonomy_fixed <- uncollapsed_taxonomy_cytb %>%
#   mutate(qseqid = str_replace_all(qseqid, "ASV", "Seq"))

# join data frames
uncollapsed_poolseq_tax_df <- asv_long_tbl %>%
  left_join(., uncollapsed_taxonomy_cytb, by = c("ASV" = "qseqid"))

# collapse these based on the appropriate taxonomic level
collapsed_poolseq_tax_df <- uncollapsed_poolseq_tax_df %>%
  select(sample.PCR, sample, rep, ASV, count, taxon, taxonomic_level) %>% ## I need to go back and fix the missing % for the species-level hits
  unique()


```


Lots of ASVs with no blast hits - how many?

```{r}
collapsed_poolseq_tax_df %>% filter(is.na(taxon)) %>%
  select(ASV) %>%
  unique()
```
86 ASVs with taxonomic info and 7,034 ASVs without.


Okay, and that leaves how many with taxonomic info?
```{r}
collapsed_poolseq_tax_df %>% 
  filter(!is.na(taxon)) %>%
  select(ASV, taxon) %>%
  unique()
  
```
That data frame still contains a bunch of NAs that didn't have blast hits. Let's go ahead and remove them.

```{r remove-NAs}
poolseq_tax_tidydf <- collapsed_poolseq_tax_df %>% 
  filter(!is.na(taxon)) %>% # remove NAs
  filter(count > 0, taxon != "Actinopteri", taxon != "Eukaryota")  %>% # not informative
  select(-ASV) # remove entries with zero reads for that sequence

# take a quick look at sample composition
poolseq_tax_tidydf %>%
  group_by(sample.PCR) %>%
  filter(!str_detect(sample, "e")) %>%
  ggplot(aes(x = sample, y = count, fill = taxon)) +
  geom_bar(stat="identity") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90),
    axis.title.x = element_text(margin = margin(t=10)),
    axis.title.y = element_text(margin = margin(r=10))
  ) +
  labs(
    y = "reads",
    x = "PCR controls"
  )
  
ggsave("pdf_outputs/CytBpositiveControls.pdf")

```

That is very interesting: The sockeye positive controls seem to have a small but ~consistent proportion of the sequencing reads assigned to pink and coho.

Two of the Chinook samples look like duds?
Two of the Coho samples are actually pink salmon?

Okay, some strange elements there... let's move on to the eDNA field samples.

## Add metadata to that data frame

```{r read-in-metadata}
# metadata tidied and ready to go
meta <- read.csv("../metadata/amalga_clean_metadata.csv") %>%
  # fix the 100m samples, which are actually between the pens and the creek
  mutate(distance = ifelse(distance == 100, -100, distance)) %>%
  mutate(dist_grp = distance)

meta
```

```{r add-meta-data}
filtered_df_w_metadata <- poolseq_tax_tidydf %>%
  left_join(., meta)

# filtered_df_w_metadata %>%
#   filter(distance > -10)%>%
#   mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>% # rename labels for plot
#   filter(!taxon %in% c("Carassius", "Oncorhynchus", "Oligocottus_maculosus", "Thalassiosira_nordenskioeldii")) %>%
#   group_by(sample.PCR) %>%
#   filter(!is.na(tide)) %>%
#   ggplot(aes(x = distance, y = count, fill = reorder(taxon, count))) +
#   geom_bar(stat="identity") +
#   facet_grid(rows = vars(tide)) +
#   theme_bw() +
#   theme(
#     axis.text.x = element_text(angle = 90),
#     axis.title.x = element_text(margin = margin(t=10)),
#     axis.title.y = element_text(margin = margin(r=10))
#   ) +
#   labs(
#     y = "Sequencing reads (summed across replicates)",
#        x = "Distance from hatchery pens (m)",
#     fill = "Taxon"
#   ) +
#   scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue", "darkviolet", 'hotpink4', 'goldenrod2', "firebrick1", 'darksalmon'))

#ggsave("pdf_outputs/CytB_taxon_counts_v_distance.pdf", width = 8, height = 5)
```

```{r}
filtered_df_w_metadata %>%
  #filter(taxon == "Carassius") %>%
  filter(str_detect(sample.PCR, "e0")) %>%
  group_by(sample.PCR, taxon) %>%
  mutate(taxon_count = sum(count)) %>%
  select(-count) %>%
  unique() %>%
  group_by(sample.PCR) %>%
  mutate(tot_count = sum(taxon_count)) %>%
  mutate(taxon_prop = taxon_count/tot_count) %>%
  ggplot(aes(x = sample.PCR, y = taxon_prop, fill = taxon)) +
  geom_bar(stat = "identity")
```
## USE THIS DATAFRAME
```{r}
# make a dataframe that includes only a single entry for each sample distance/tide/taxon with the mean read count
# multiple lines with the mean will mean that the mean is double (or more)-counted!
df_taxa_uniqued <- filtered_df_w_metadata %>%
  filter(str_detect(sample.PCR, "e0")) %>%
  group_by(sample.PCR, taxon) %>%
  mutate(taxon_count = sum(count)) %>%
  select(-count) %>%
  unique()

tide.dist.unique.taxa <- df_taxa_uniqued %>%
    filter(distance > -10)%>%
    mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>% # rename labels for plot
    filter(!taxon %in% c("Carassius", "Oncorhynchus", "Oligocottus_maculosus", "Thalassiosira_nordenskioeldii")) %>%
  filter(!is.na(tide)) %>%
    group_by(distance, taxon, tide) %>%
  mutate(mean_reads = mean(taxon_count)) %>%
  select(taxon, mean_reads, type, tide, distance) %>%
  unique()
```


```{r make-plot}
tide.dist.unique.taxa %>% 
  ggplot(aes(x = distance, y = mean_reads, fill = reorder(taxon, mean_reads))) +
  geom_bar(stat="identity") +
  facet_grid(rows = vars(tide), scales = "free_y") +
  theme_bw() +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12, face = "italic"),
    legend.title = element_blank(),
    strip.text = element_text(size = 12)
    ) +
  labs(
    y = "Sequencing reads (mean across replicates)",
       x = "Distance from hatchery pens (m)"  )+
  scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue", "darkviolet", 'hotpink4', 'goldenrod2', "firebrick1", 'darksalmon'))

#ggsave("pdf_outputs/CytB_mean_reads_v_distance.pdf", width = 9, height = 5)
```
One thing here: there was definitely a plate-affect in the sequencing depth, but because the samples and replicates were distributed randomly, the signal of much higher read depth on the outgoing tide is probably real.

Mean number of reads per replicate (rather than sum)
```{r plot-mean-reads}
# mean reads, grouped by distance
# filtered_df_w_metadata %>%
#   filter(count >0) %>%
#     filter(distance > -10)%>%
#     mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>% # rename labels for plot
#     filter(!taxon %in% c("Carassius", "Oncorhynchus", "Oligocottus_maculosus", "Thalassiosira_nordenskioeldii")) %>%
#   filter(!is.na(tide)) %>%
#     group_by(distance, taxon, tide) %>%
#   mutate(mean_reads = mean(count)) %>%
#   ggplot(aes(x = distance, y = mean_reads, fill = reorder(taxon, mean_reads))) +
#   geom_bar(stat="identity") +
#   facet_grid(rows = vars(tide), scales = "free_y") +
#   theme_bw() +
#   theme(
#     axis.text.x = element_text(size = 14),
#     axis.text.y = element_text(size = 14),
#     axis.title.x = element_text(margin = margin(t=10), size = 14),
#     axis.title.y = element_text(margin = margin(r=10), size = 14),
#     legend.text = element_text(size = 12),
#     legend.title = element_blank(),
#     strip.text = element_text(size = 12)
#     ) +
#   labs(
#     y = "Sequencing reads (mean across replicates)",
#        x = "Distance from hatchery pens (m)"  )+
#   scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue", "darkviolet", 'hotpink4', 'goldenrod2', "firebrick1", 'darksalmon'))
#   
```
Out of curiosity, how does the qPCR compare?
```{r}
# read in qPCR data
qpcr <- read_csv("csv_outputs/qPCRcleanDataFrame.csv")
```

```{r}
qpcr_select <- qpcr %>%
  select(sample, rep_mean, location, distance, Sample_type, tide) %>%
  filter(distance > -10)%>%
    mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) # rename labels for plot
```

  
```{r}
chum_mean_ct_df <- filtered_df_w_metadata %>%
  filter(count >0) %>%
    filter(distance > -10)%>%
    mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>% # rename labels for plot
    #filter(taxon %in% c("Oncorhynchus_keta")) %>%
  filter(!is.na(tide)) %>%
  group_by(distance, taxon, tide) %>%
  mutate(mean_reads = mean(count)) %>%
  select(taxon, mean_reads, type, tide, distance) %>%
  unique() %>%
  ungroup()


joined_df <- chum_mean_ct_df %>%
  left_join(., qpcr_select, by = c("distance", "tide"))
```


```{r}
ggplot() +
  geom_bar(data = chum_mean_ct_df, aes(x = distance, y = mean_reads, fill = taxon), stat="identity") +
  #geom_point(data = qpcr_select, aes(x = distance, y = rep_mean), color = "firebrick2") +
  facet_grid(rows = vars(tide), scales = "free_y") +
  theme_bw() +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12),
    ) +
  labs(
    y = "Sequencing reads (mean across replicates)",
       x = "Distance from hatchery pens (m)")

#ggsave("pdf_outputs/CytBketaReadsVdistance.pdf", width = 9, height = 5)
```
Why does that one look different from the other?? There's data duplication going on somewhere.

That doesn't work because of the different y scales, but I could layer the plots together?

```{r}
ggplot() +
  #geom_bar(data = chum_mean_ct_df, aes(x = distance, y = mean_reads), fill = "darksalmon", stat="identity") +
  geom_point(data = qpcr_select, aes(x = distance, y = log(rep_mean)), color = "firebrick2") +
  facet_grid(rows = vars(tide), scales = "free_y") +
  theme_bw() +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12),
    ) +
  labs(
    y = "Mean DNA copy number (log10)",
       x = "Distance from hatchery pens (m)")


ggsave("pdf_outputs/qPCRcountsVdistance.pdf", width = 9, height = 5)

```
  
## Make taxon proportions plot

```{r proportions-for-plotting}
prop.df <- filtered_df_w_metadata %>% 
  filter(type == "eDNA") %>%
      filter(!taxon %in% c("Carassius", "Oligocottus_maculosus", "Thalassiosira_nordenskioeldii")) %>% #these are minor constituents that include the postive control and a scuplin
  select(-taxonomic_level, -location, -label, -dist_grp) %>%
  filter(!is.na(tide)) %>%
  #filter(sample %in% c("e00050", "e00051")) %>%
  group_by(distance, tide) %>% # group by distance
  mutate(total_reads = sum(count)) %>%
  group_by(distance, tide, taxon) %>%
  mutate(taxon_count = sum(count)) %>%
  select(-count, -rep, -sample.PCR, -sample) %>%
  unique() %>% # if we don't call unique here, it duplicates the summed entries
  mutate(taxon_prop = taxon_count/total_reads)


```


```{r proportions-plot}
prop.df %>%
  filter(distance > -10)%>%
  mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>%
  ggplot(aes(distance, taxon_prop, fill = reorder(taxon, taxon_prop)))+
  geom_bar(stat="identity", position="stack") +
  facet_grid(rows = vars(tide)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12, face = "italic"),
    strip.text = element_text(size = 12),
    legend.title = element_blank()
    ) +
  labs(y = "Proportion of reads",
       x = "Distance from hatchery pens (m)",
       fill = "Taxon") +
   scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue", "darkviolet", 'hotpink4', 'goldenrod2', "firebrick1",'darksalmon'))


```

Look at the same with a large proportion of coho

```{r}
prop.df %>%
  filter(taxon == "Oncorhynchus_kisutch" & taxon_prop >0.5)

```

Does that show up in all of the replicates?
```{r}
# Looks like an issue of low read depth in all other replicates
# but coho doesn't show up in any of the other replicates...
filtered_df_w_metadata %>%
  filter(tide == "outgoing", distance == "1680")

```
I need to look more closely at the read depth variation because that can cause these one-offs to look artificially inflated.


```{r}
filtered_df_w_metadata
  
```


```{r rep-viz-fct}
# wrap up that visualization tool into a function
plot_reps <- function(dist){
filtered_df_w_metadata %>%
  filter(distance == dist) %>%
  ggplot(aes(sample.PCR, count, fill = reorder(taxon, count)))+
  geom_bar(stat="identity", position="stack") +
  facet_grid(rows = vars(tide)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle =90),
    axis.title.x = element_text(margin = margin(t=10)),
    axis.title.y = element_text(margin = margin(r=10))
  ) +
  labs(y = "Number of reads",
       x = "sample PCR replicate",
       fill = "Taxon")

}

```

```{r}
# make a list of the distances
nums <- filtered_df_w_metadata %>%
  ungroup() %>%
  select(distance) %>%
  filter(!is.na(distance)) %>%
  unique() %>%
  arrange(distance)

dists <- nums$distance
```

```{r}
# loop over those to take a quick look
lapply(dists, plot_reps)
```

It seems like, if there's a lot of water movement, and relatively little DNA, then there's some higher probability that patchily distributed species might be detected as a spike when present. 


Look at controls:
```{r}
filtered_df_w_metadata %>%
  filter(str_detect(sample, "NEG")) %>%
   ggplot(aes(sample.PCR, count, fill = reorder(taxon, count)))+
  geom_bar(stat="identity", position="stack") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle =90),
    axis.title.x = element_text(margin = margin(t=10)),
    axis.title.y = element_text(margin = margin(r=10))
  ) +
  labs(y = "Number of reads",
       x = "sample PCR replicate",
       fill = "Taxon")
  
```

These negative controls are super clean (only two with any reads: 1 with 1 read and the other with 3 reads)

```{r}
filtered_df_w_metadata %>%
  filter(type != "eDNA") %>%
   ggplot(aes(sample.PCR, count, fill = reorder(taxon, count)))+
  geom_bar(stat="identity", position="stack") +
  facet_grid(rows = vars(type)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle =90),
    axis.title.x = element_text(margin = margin(t=10)),
    axis.title.y = element_text(margin = margin(r=10))
  ) +
  labs(y = "Number of reads",
       x = "sample PCR replicate",
       fill = "Taxon")
  
ggsave("pdf_outputs/cytb_blanks.pdf")
```


## Presence/absence

```{r PA-plot}

prop.df %>%
    filter(distance > -10) %>%
  mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>%
  ggplot(aes(distance, fill = reorder(taxon, taxon_prop)))+
  geom_bar(stat="count", position="stack") +
  facet_grid(rows = vars(tide)) +
  theme_bw() +
  theme(
    axis.title.x = element_text(margin = margin(t=10)),
    axis.title.y = element_text(margin = margin(r=10))
  ) +
  labs(y = "Number of taxa",
       x = "Distance from hatchery pens (m)",
       fill = "Taxon") +
   scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue", "darkviolet", 'hotpink4', 'goldenrod2', "firebrick1", 'darksalmon'))

```

## summary stuff

```{r}
asv_long_tbl %>%
  group_by(sample.PCR, sample) %>%
  summarise(tot_count = sum(count)) %>%
  filter(str_detect(sample, "e"))

# it's like the C reps didn't work?
```

```{r}
asv_long_tbl %>% 
  filter(str_detect(sample, "e"),
         count>0) %>%
  left_join(., meta, by = "sample") %>%
  ggplot(aes(x = sample, y = count)) +
  geom_point() +
  facet_grid(rows = vars(rep), scales = "free_y") +
  theme(
    axis.text.x = element_text(angle = 90)
  )

```
Like a bunch of the C samples didn't work.


```{r}
test.df <- meta %>% 
  left_join(., asv_long_tbl, by = "sample") %>%
  filter(str_detect(sample, "e")) %>%
  group_by(sample, tide, distance, rep) %>%
  mutate(sum_ct = sum(count))
  
test.df %>%
  ggplot(aes(x = distance, y = sum_ct)) +
  geom_point() +
  facet_grid(rows = vars(rep), scales = "free_y")
```

