---
title: "salmonid-cytb-analysis"
output: html_notebook
---

9 March 2022

Things got a bit chaotic in the `09-salmon-cytb-taxonomic-filtering.Rmd` doc, so the hope is that this one can be analyses where the taxonomy is already done.


```{r load-libraries}
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)
library(readxl)
library(readr)

```


```{r load-asv-table}
## Read in ASV-sample table for Poolseq data

asv_tbl <- read.csv("csv_outputs/CytBsampleTable.csv")

# rename the samples because the sample sheet for the MiSeq wasn't accurate (bec of the extra positive controls for salmon species)
# read in the renaming table
rename_plate1 <- read_xlsx("../cytb/Salmon CytB_Oct2021_updated.xlsx", sheet = "ctyb_plate1_samplesheet")
rename_plate2 <- read_xlsx("../cytb/Salmon CytB_Oct2021_updated.xlsx", sheet = "cytb_plate2_samplesheet")

rename_all <- bind_rows(rename_plate1, rename_plate2) %>%
  mutate(sample = str_replace_all(sample, "_", "-"))

asv_renamed_tbl <- asv_tbl %>%
  left_join(., rename_all, by = c("X"="sample")) 

asv_renamed2 <- asv_renamed_tbl %>%
  select(real_sample_name, X, 2:length(asv_renamed_tbl)) %>%
  select(-X) %>%
   rename(sample.PCR = real_sample_name) %>%
    mutate(sample.PCR = str_replace_all(sample.PCR, fixed(" "), "")) %>%
    separate(sample.PCR, into = c("sample", "rep", "extra"), remove = F) 

# pivot longer for ease of joining
asv_long_tbl <- asv_renamed2 %>%
  pivot_longer(cols = 5:length(asv_renamed2), names_to = "ASV", values_to = "count") %>%
  filter(count>0)

```

How many total ASVs? (7120)
```{r}
asv_long_tbl %>%
  select(ASV) %>%
  unique()
```

How many reads per sample?
```{r}
asv_long_tbl %>%
  group_by(sample.PCR) %>%
  summarise(sum(count))
```
The third plate of replicates really sequenced terribly! Thousands of reads for the other samples and 2-3 reads for the "C" replicates. I might need to drop those all together?


How many reads per sample?
```{r}
asv_long_tbl %>%
  group_by(sample) %>%
  mutate(tot_reads = sum(count)) %>%
  group_by(sample.PCR) %>%
  mutate(rep_reads = sum(count)) %>%
  select(sample.PCR, sample, rep, tot_reads, rep_reads) %>%
  unique() %>% # to avoid duplicate entries
  mutate(rep_prop = rep_reads/tot_reads) %>%
  filter(rep_prop < 0.05) %>%
  arrange(rep)
```
Still a bunch from all of the plates - so better to use the same approach as with 12S where I remove the low read depth replicates.


```{r}
asv_long_tbl %>%
  group_by(sample) %>%
  mutate(tot_reads = sum(count)) %>%
  group_by(sample.PCR) %>%
  mutate(rep_reads = sum(count)) %>%
  select(sample.PCR, sample, rep, tot_reads, rep_reads) %>%
  unique() %>% # to avoid duplicate entries
  mutate(rep_prop = rep_reads/tot_reads) %>%
  filter(rep_prop < 0.2) %>%
  ggplot(aes(x = rep_prop)) +
  geom_density()

```
Somewhere between 5-10% again makes sense for Cyt b. I'll keep it consistent with the 12S data for simplicity and use 6%.


```{r}
reps.to.remove <- asv_long_tbl %>%
  group_by(sample) %>%
  mutate(tot_reads = sum(count)) %>%
  group_by(sample.PCR) %>%
  mutate(rep_reads = sum(count)) %>%
  select(sample.PCR, sample, rep, tot_reads, rep_reads) %>%
  unique() %>% # to avoid duplicate entries
  mutate(rep_prop = rep_reads/tot_reads) %>%
  filter(rep_prop < 0.06,
         !sample %in% c("POSITIVE", "NEGATIVE")) %>%
  arrange(rep)
  

```
No obvious plate affect since replicates from all three plates (A, B, C) are present, but plate C has a notably larger number of low-read samples to toss. I'm not going to filter the negative and positive controls because those serve a different purpose here.

72 PCR replicates are removed due to low read depth (<6% of the total reads for all three replicates for that sample).


```{r}
# use an anti-join to remove those replicates from the asv dataset
asv.tbl.good.reps <- asv_long_tbl %>%
  anti_join(., reps.to.remove)

```

Now use this filtered dataframe for the analysis.


# Merge taxonomy with ASV table

```{r load-taxonomy-df}
taxonomy <- read_csv("csv_outputs/salmonidCytbTaxonomyDataframe.csv")

# a few specific easy fixes:
clean_taxonomy <- taxonomy %>%
  mutate(taxon = ifelse(qseqid == "ASV3192", "Oncorhynchus_mykiss", taxon)) %>% # the others are out of range
    mutate(taxonomic_level = ifelse(qseqid == "ASV3192", "species", taxonomic_level)) 

  
```
The only ASVs that didn't get to species were Carassius (goldfish) and Salvelinus.

Combine taxonomy and sample ASV table
```{r merge-taxonomy-and-sample-tbl}
# join data frames
taxonomy.asv.df <- asv.tbl.good.reps %>%
  left_join(., clean_taxonomy, by = c("ASV" = "qseqid"))

# collapse these based on the appropriate taxonomic level
collapsed.tax.asv.df <- taxonomy.asv.df %>%
  select(sample.PCR, sample, rep, ASV, count, taxon, taxonomic_level) %>% 
  unique() %>%
  filter(!is.na(taxon)) %>% # remove ASVs with no taxonomy
  group_by(sample.PCR, taxon) %>% # collapse multiple ASVs into single taxa by summing across their read counts
  mutate(taxon_count = sum(count)) %>%
  select(-count, -ASV) %>% # the taxon_count variable now takes the place of "count"
  unique()

```

There should only be one entry per sample per taxon 
```{r}
collapsed.tax.asv.df %>%
  tally() %>%
  filter(n >1) # there are no instances where the same taxon is included in multiple lines for a given sample. Great.
  
```
## Add metadata to that data frame

```{r read-in-metadata}
# metadata tidied and ready to go
meta <- read.csv("../metadata/amalga_clean_metadata.csv") %>%
  # fix the 100m samples, which are actually between the pens and the creek
  mutate(distance = ifelse(distance == 100, -100, distance)) %>%
  mutate(dist_grp = distance)

meta
```

Combine metadata with ASV taxonomy table
```{r add-metadata-to-asv-tax-tbl}
# combine metadata
tax.w.meta <- collapsed.tax.asv.df %>%
  left_join(., meta, by = "sample") %>%
  filter(!is.na(location)) # remove the controls with no metadata

# create a data frame with the mean read counts per taxon/distance/tide
slim.df.to.plot <- tax.w.meta %>%
  group_by(taxon, tide, distance) %>%
  mutate(mean_reads = mean(taxon_count)) %>% # this is the mean number of reads per taxon/tide/distance
  select(taxon, tide, distance, mean_reads) %>% # need to remove this to avoid counting multiple times
  unique()

```

```{r}
tax.w.meta %>%
  filter(distance == 2000)%>%
  arrange(sample)

```

## Plot mean reads across transect distance


```{r plot-mean-reads-per-taxon-distance-tide}
slim.df.to.plot %>%    
  filter(!taxon %in% c("Carassius", "Oligocottus_maculosus", "Thalassiosira_nordenskioeldii"),
         distance>-10) %>%
    mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>% # rename labels for plot
  ggplot(aes(x = distance, y = mean_reads, fill = reorder(taxon, mean_reads))) +
  geom_bar(stat = "identity") +
  facet_grid(rows = vars(tide), scales = "free_y") +
  theme_bw() +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12, face = "italic"),
    legend.title = element_blank(),
    strip.text = element_text(size = 12)
    ) +
  labs(
    y = "Sequencing reads (mean across replicates)",
       x = "Distance from hatchery pens (m)"  ) +
  scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue", "darkviolet", 'hotpink4', 'goldenrod2', "firebrick1", 'darksalmon'))

#ggsave("pdf_outputs/CytB_mean_reads_v_distance.pdf", width = 9, height = 5)
```
## Make taxon proportions plot

```{r proportions-for-plotting}
prop.df <- tax.w.meta %>%
  filter(!is.na(tide),
         type == "eDNA", # remove the controls with no metadata
        !taxon %in% c("Carassius", "Oligocottus_maculosus", "Thalassiosira_nordenskioeldii")) %>% #these are minor constituents that include the postive control and a scuplin
  select(-taxonomic_level, -location, -label, -dist_grp) %>%
  #filter(sample %in% c("e00050", "e00051")) %>%
  group_by(distance, tide, taxon) %>% # group by distance
  mutate(tot_tax_reads = sum(taxon_count)) %>%
  #group_by(distance, tide, taxon) %>%
  select(distance, taxon, tide, tot_tax_reads) %>%
  unique() %>% # if we don't call unique here, it duplicates the summed entries
  group_by(distance, tide) %>%
  mutate(taxon_prop = tot_tax_reads/sum(tot_tax_reads))


```


```{r proportions-plot}
prop.df %>%
  filter(distance > -10)%>%
  mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>%
  ggplot(aes(distance, taxon_prop, fill = reorder(taxon, taxon_prop)))+
  geom_bar(stat="identity", position="stack") +
  facet_grid(rows = vars(tide)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12, face = "italic"),
    strip.text = element_text(size = 12),
    legend.title = element_blank()
    ) +
  labs(y = "Proportion of reads",
       x = "Distance from hatchery pens (m)",
       fill = "Taxon") +
   scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue",  'hotpink4', "firebrick1", "darkviolet", 'goldenrod2', 'darksalmon'))


#ggsave("pdf_outputs/CytB_taxon_proportions_v_distance.pdf", width = 9, height = 5)

```

## Presence/absence

```{r PA-plot}
prop.df %>%
    filter(distance > -10) %>%
  mutate(tide = ifelse(tide == "incoming", "incoming tide (AM)", "outgoing tide (PM)")) %>%
  ggplot(aes(distance, fill = reorder(taxon, taxon_prop)))+
  geom_bar(stat="count", position="stack") +
  facet_grid(rows = vars(tide)) +
  theme_bw() +
 theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(margin = margin(t=10), size = 14),
    axis.title.y = element_text(margin = margin(r=10), size = 14),
    legend.text = element_text(size = 12, face = "italic"),
    strip.text = element_text(size = 12),
    legend.title = element_blank()
    ) +
  labs(y = "Number of taxa present",
       x = "Distance from hatchery pens (m)",
       fill = "Taxon") +
   scale_fill_manual(values = c('slateblue2', 'aquamarine3', "dodgerblue",  'hotpink4', "firebrick1", "darkviolet", 'goldenrod2', 'darksalmon'))



#ggsave("pdf_outputs/CytB_taxonPA_v_distance.pdf", width = 9, height = 5)
```


## Fitting a line to compare with qPCR


Get the counts for chum salmon across the transect and figure out how reads counts compare to the qPCR results.


```{r}
cytb.linearModel.plot <- slim.df.to.plot %>% 
  filter(taxon == "Oncorhynchus_keta",
         distance>-10) %>%
    mutate(tide = ifelse(tide == "incoming", "Incoming tide (AM)", "Outgoing tide (PM)")) %>% # rename labels for plot
    ggplot(aes(x = distance, y = log10(mean_reads), color = tide, shape = tide)) +
    geom_point() +
    geom_smooth(method = "lm", alpha = 0.15) +
    #facet_grid(rows = vars(tide), scales = "free_y") +
    theme_minimal() +
      theme(
    #axis.title.x = element_text(margin = margin(t=12), size = 10),
    axis.title.x = element_blank(),
    axis.title.y = element_text(margin = margin(r=12), size = 10),
    axis.text = element_text(size = 10),
    legend.title = element_blank(),
    #legend.title = element_text(size = 14),
    legend.text = element_text(size = 10),
    #legend.position = c(0.8, 0.85)
    legend.position = "none"
  ) +
    labs(
      y = "Mean reads (log10)",
         x = "Distance from hatchery pens (m)",
      title = "Cyt B salmonid metabarcoding: chum salmon") +
  scale_color_manual(values = c("firebrick2", "dodgerblue"))

cytb.linearModel.plot

ggsave("pdf_outputs/cytbLog10linearModelChum_long.pdf", width = 6, height = 2)

```

Linear model using log10 
```{r}
# create a new variable with the log10 mean reads
lm.df <- slim.df.to.plot %>% 
  filter(taxon == "Oncorhynchus_keta",
         distance>-10) %>%
  mutate(log_reads = log10(mean_reads))

# incoming tide
incoming_df <- lm.df %>%
  filter(tide == "incoming")

# outgoing tide
outgoing_df <- lm.df %>%
  filter(tide == "outgoing")

# linear model
incoming_model<-lm(log_reads~distance,data=incoming_df)

summary(incoming_model)$r.squared

outgoing_model <- lm(log_reads~distance, data=outgoing_df)

summary(outgoing_model)$r.squared

```
R^2
incoming = 0.65
outgoing = 0.73

I bet the fit is even better until 1000 m.

```{r}
# chop off the dataframe after 1000 m
# incoming tide
incoming_df <- lm.df %>%
  filter(tide == "incoming", 
         distance <1000)

# outgoing tide
outgoing_df <- lm.df %>%
  filter(tide == "outgoing",
         distance<1000)

# linear model
incoming_model<-lm(log_reads~distance,data=incoming_df)

summary(incoming_model)$r.squared

outgoing_model <- lm(log_reads~distance, data=outgoing_df)

summary(outgoing_model)$r.squared

```
That does not improve the relationship for incoming and only slightly for outgoing.
incoming = 0.24
outgoing = 0.77

